WikiWatershed/mmw-geoprocessing
# Subbasin MultiOperation Benchmark

## Overview

Adds two benchmarking scripts, `benchmark-mapshed` and `benchmark-subbasin`, that run MapShed and Subbasin the "naive" way, using the existing `/run` endpoint. Also updates existing benchmark script to run MapShed and Subbasin the new way, using the `/multi` endpoint. Adds a `Pipenv` file for managing the Python dependencies for the benchmark files, and updates the `setup` script to install those dependencies. Adds a container `benchmark` script that runs all three sub-scripts given flags.

These benchmarks run against `http://localhost:8090/`. When running `./scripts/server` in this project, the benchmarks run against a server running natively on the developer's machine. When running against MMW, the benchmarks run against a server in the Worker VM.

These are some graphs from benchmarks comparing these cases across five runs:

![image](https://user-images.githubusercontent.com/1430060/37616651-05c72b8e-2b87-11e8-9c2b-7f7efe82c2a1.png)

![image](https://user-images.githubusercontent.com/1430060/37616678-14db14fa-2b87-11e8-8307-54fedabc0d3b.png)

In both cases, the blue lines are of the current implementation and the orange lines of the new one. The solid lines are native tests, and the dashed lines are in the Worker VM. Notably, Subbasin did not finish in the Worker VM at all.

The difference between native and Worker VM runs is due to resource allocation: the Worker VM has only 2 vCPUs and only 2 GB RAM, whereas my native machine has a quadcore CPU with 16 GB of RAM. Staging and Production environments have beefier Worker VMs, and have correspondingly shorter runtimes. This process seems to be more CPU bound than RAM bound.

Connects #83 

### Demo

<details>
<summary>

```
$ ./scripts/benchmark --sync
```

</summary>

```
Timing RasterGroupedAverage ->

RasterGroupedAverage, run 1 -> 1.310392 s
RasterGroupedAverage, run 2 -> 0.666542 s
RasterGroupedAverage, run 3 -> 0.455111 s
RasterGroupedAverage, run 4 -> 0.240777 s
RasterGroupedAverage, run 5 -> 3.789703 s

RasterGroupedAverage average -> 1.2925049999999998 s

Timing HUC8 RasterLinesJoin ->

HUC8 RasterLinesJoin, run 1 -> 1.282524 s
HUC8 RasterLinesJoin, run 2 -> 1.219903 s
HUC8 RasterLinesJoin, run 3 -> 1.346176 s
HUC8 RasterLinesJoin, run 4 -> 1.509372 s
HUC8 RasterLinesJoin, run 5 -> 1.572968 s

HUC8 RasterLinesJoin average -> 1.3861885999999999 s

Timing HUC8 RasterSummary ->

HUC8 RasterSummary, run 1 -> 0.232893 s
HUC8 RasterSummary, run 2 -> 0.465685 s
HUC8 RasterSummary, run 3 -> 0.164735 s
HUC8 RasterSummary, run 4 -> 2.80515 s
HUC8 RasterSummary, run 5 -> 0.60106 s

HUC8 RasterSummary average -> 0.8539046000000001 s
```

</details>

<details>
<summary>

```
$ ./scripts/benchmark --mapshed
```

</summary>

```
Timing HUC8 MultiOperation MapShed ->

HUC8 MultiOperation MapShed, run 1 -> 13.442911 s
HUC8 MultiOperation MapShed, run 2 -> 6.674352 s
HUC8 MultiOperation MapShed, run 3 -> 6.212541 s
HUC8 MultiOperation MapShed, run 4 -> 6.86085 s
HUC8 MultiOperation MapShed, run 5 -> 6.103041 s

HUC8 MultiOperation MapShed average -> 7.858739 s

Timing HUC8 MapShed, separate requests per operation ->

HUC8 MapShed, 7 operations, run 1 -> 8.18 s
HUC8 MapShed, 7 operations, run 2 -> 8.30 s
HUC8 MapShed, 7 operations, run 3 -> 8.03 s
HUC8 MapShed, 7 operations, run 4 -> 10.61 s
HUC8 MapShed, 7 operations, run 5 -> 9.56 s
```
</details>
<details>
<summary>

```
$ ./scripts/benchmark --subbasin
```

</summary>

```
Timing 61 HUC12 MultiOperation Subbasin ->

61 HUC12 MultiOperation Subbasin, run 1 -> 12.072989 s
61 HUC12 MultiOperation Subbasin, run 2 -> 12.301599 s
61 HUC12 MultiOperation Subbasin, run 3 -> 12.132871 s
61 HUC12 MultiOperation Subbasin, run 4 -> 12.51495 s
61 HUC12 MultiOperation Subbasin, run 5 -> 13.357691 s

61 HUC12 MultiOperation Subbasin average -> 12.47602 s

Timing 61 HUC12 Subbasin, separate requests per operation ->

61 HUC12 Subbasin, 7 operations, run 1 -> 78.48 s
61 HUC12 Subbasin, 7 operations, run 2 -> 92.60 s
61 HUC12 Subbasin, 7 operations, run 3 -> 75.59 s
61 HUC12 Subbasin, 7 operations, run 4 -> 59.18 s
61 HUC12 Subbasin, 7 operations, run 5 -> 67.95 s
```

</details>

### Notes

The new `pipenv` dependency should added to the README, but it is far too out of date, so I decided to only add the message in the `setup` script. #56 exists to update the README.

The added value of `request-timeout` is to allow longer requests to complete, which take more than the default 20 s (some of the MultiOperation requests for Schuylkill size HUC-8s take 45+ seconds).

## Testing Instructions

* Check out this branch and run `./scripts/setup`
    * If you see an error message about [`pipenv`](https://docs.pipenv.org/), install it, and run `./scripts/setup` again
* Run the server with `./scripts/server`
* Run the old benchmarks with `./scripts/benchmark --sync`
* Run the old (multiple requests to `/run`) and the new (single request to `/multi`) style MapShed benchmarks with `./scripts/benchmark --mapshed`
* Run the old and the new style Subbasin benchmarks with `./scripts/benchmark --subbasin`
